### PR TITLE
*: Simplify build steps

### PR DESCRIPTION
**Pull request by Xuanwo**
## Changes

- Use go.mod to manage go related deps, pin gogo/protobuf to tag v0.5
- Copy proto files from gogo/protobuf to this repo
- Remove not needed shell scripts

## Results

After this PR get merged, we don't need to checkout gogo/protobuf anymore

Close https://github.com/pingcap/tipb/pull/111

Signed-off-by: Xuanwo <github@xuanwo.io>

### CODE REVIEW

**Comment by breeswish** [comment]
@tiancaiamao Could you take a look?

**Comment by tiancaiamao** [comment]
LGTM

**Comment by Xuanwo** [comment]
ping @breeswish @tiancaiamao What else do I need to do to make this PR mergeable?

**Comment by Xuanwo** [comment]
@winoros Please take a look on this PR

**Comment by breeswish** [comment]
I have a question. Looks like gogo/protobuf source files are placed in this repo. Is it necessary?

**Comment by Xuanwo** [comment]
Yes, protoc relies on these definition issue files when generating code. Whereas before we chose to clone the entire repo and then git checkout it, now we put it directly in the project and update these files when and only when we decide to update the dependencies.

**Comment by breeswish** [comment]
Is it possible to use go mod to maintain these dependencies, and exclude them from the repository source code?

**Comment by Xuanwo** [comment]
ping @breeswish, The perfect version is here:

- All gogo/google proto files are maintainer via go mod.
- No include for rust any more, use the same proto files.
- Migrate to github actions, make sure all targets are tested and faster, take https://github.com/Xuanwo/tipb/runs/1569015747?check_suite_focus=true as a look.

File: Makefile
Comment (breeswish) [inline Makefile:1]: Is it possible to use `go list -f "{{.Dir}}" -m github.com/gogo/protobuf`?

**Review by breeswish — COMMENTED** [review]
Mostly LGTM, thanks!

File: Makefile
Comment (Xuanwo) [reply Makefile:1]: Wow, I think it works.

**Comment by Xuanwo** [comment]
@breeswish implemented, PTAL

File: build.rs
Comment (breeswish) [inline build.rs:8]: I guess this might cause problems when tipb is included by TiKV, due to this variable only exists in tipb's Makefile. How about referring to how kvproto does?

Comment (Xuanwo) [reply build.rs:8]: They need an `include` which stores all protos needed.

Maybe I should revert the changes to `include`, or find a way to embed the files.

Comment (breeswish) [reply build.rs:8]: Got it, looks like we should stick to how [kvproto](https://github.com/pingcap/kvproto/tree/master/include) does.

**Comment by Xuanwo** [comment]
PTAL @breeswish

File: Makefile
Comment (breeswish) [inline Makefile:11]: Now that we have protobuf definitions in the repository, do we still need GOGO_PROTOBUF?

**Comment by Xuanwo** [comment]
ping @breeswish , github lost your comments.

> Now that we have protobuf definitions in the repository, do we still need GOGO_PROTOBUF?

Yes.

`protoc-gen-gofast` requires a special set of protos called `gogoproto` which included in `gogo/protobuf`.

**Comment by Xuanwo** [comment]
ping @breeswish , PTAL

**Comment by breeswish** [comment]
Thanks for the contribution, I will try with this PR this weekend!

**Comment by Xuanwo** [comment]
@breeswish Hello, any feedback on this PR?

File: go.mod
Comment (breeswish) [inline go.mod:5]: I think we'd better stick to the same version requirement as https://github.com/pingcap/kvproto/blob/master/go.mod

**Comment by breeswish** [comment]
Currently TiDB compilation will fail when using your PR:

```
go: finding module for package google.golang.org/grpc/naming
go: finding module for package google.golang.org/grpc/examples/helloworld/helloworld
go: downloading honnef.co/go/tools v0.1.0
go: downloading google.golang.org/grpc/examples v0.0.0-20201226181154-53788aa5dcb4
go: found google.golang.org/grpc/examples/helloworld/helloworld in google.golang.org/grpc/examples v0.0.0-20201226181154-53788aa5dcb4
go: finding module for package google.golang.org/grpc/naming
go: downloading google.golang.org/genproto v0.0.0-20200806141610-86f49bd18e98
github.com/pingcap/tidb/ddl/util tested by
	github.com/pingcap/tidb/ddl/util.test imports
	go.etcd.io/etcd/integration imports
	go.etcd.io/etcd/proxy/grpcproxy imports
	google.golang.org/grpc/naming: module google.golang.org/grpc@latest found (v1.34.0), but does not contain package google.golang.org/grpc/naming
```

I'm not sure whether this is related to https://github.com/pingcap/tipb/pull/199#discussion_r550398634

**Comment by Xuanwo** [comment]
Let me check it out.

**Comment by Xuanwo** [comment]
I got this fixed.

**Comment by Xuanwo** [comment]
ping @breeswish to check it again.

File: go.mod
Comment (Xuanwo) [reply go.mod:5]: tipb has it's own requirements. go modules always pick the latest same major version while build, so we only need to take care don't intreduce higher versions.

Comment (breeswish) [reply go.mod:5]: Actually I'm thinking whether it is possible to make kvproto & tipb identical. I think they are the same and currently I cannot think out reasons to maintain two different approaches.

**Comment by breeswish** [comment]
Seems to be still failing..

```
$ GO111MODULE=on go mod edit -replace github.com/pingcap/tipb=github.com/Xuanwo/tipb@a0504ec                                 
$ make
CGO_ENABLED=1 GO111MODULE=on go build  -tags codes  -ldflags '-X "github.com/pingcap/parser/mysql.TiDBReleaseVersion=v4.0.0-beta.2-1941-gdad3b5e87-dirty" -X "github.com/pingcap/tidb/util/versioninfo.TiDBBuildTS=2021-01-04 04:32:29" -X "github.com/pingcap/tidb/util/versioninfo.TiDBGitHash=dad3b5e87d21caf29bd7f0c7d968cd410e7a5cdd" -X "github.com/pingcap/tidb/util/versioninfo.TiDBGitBranch=master" -X "github.com/pingcap/tidb/util/versioninfo.TiDBEdition=Community" ' -o bin/tidb-server tidb-server/main.go
go: downloading github.com/Xuanwo/tipb v0.0.0-20201231084958-a0504ec508a5
go: finding module for package google.golang.org/grpc/naming
../../../go/pkg/mod/google.golang.org/api@v0.18.0/internal/pool.go:10:2: module google.golang.org/grpc@latest found (v1.34.0), but does not contain package google.golang.org/grpc/naming
```

**Review by breeswish — APPROVED** [review]
LGTM

**Comment by breeswish** [comment]
I'm giving my approval. However I recommend you to research on whether it is possible to use the same code base for kvproto and tipb. If we can use the same code base, then the maintenance effort could be greatly reduced.

**Comment by Xuanwo** [comment]
OK, I will find a solution out.

**Comment by Xuanwo** [comment]
@breeswish What else should I do to get this PR merged?

**Comment by breeswish** [comment]
@Xuanwo You need an approval from another one.

**Comment by Xuanwo** [comment]
ping @tiancaiamao to take a look

**Comment by Xuanwo** [comment]
@winoros Do you have time to take a look on this PR?

**Review by skyzh — APPROVED** [review]
This patch looks good to me. @littlefall told me that TiFlash depends on an older version of tipb, and does not even compile with current master. Therefore, this PR won't affect current versions of TiFlash / TiSpark. I suggest merge this PR first, and later solve these issues when upgrading tipb dependency.

**Comment by Xuanwo** [comment]
Let's get this PR merged! @breeswish @skyzh

**Comment by skyzh** [comment]
> Let's get this PR merged! @breeswish @skyzh

It seems that we have to modify required checks before merging. Is it possible to make both GitHub Action and CircleCI work, therefore not to affect other PRs?

**Comment by LittleFall** [comment]
> This patch looks good to me. @LittleFall told me that TiFlash depends on an older version of tipb, and does not even compile with current master. Therefore, this PR won't affect current versions of TiFlash / TiSpark. I suggest merge this PR first, and later solve these issues when upgrading tipb dependency.

sorry, wait a minute for merging it, I'm trying further tiflash testing. And you can still solve the CI problem first.

**Comment by Xuanwo** [comment]
It's hard to migrate to CircleCI (we need to build our images for that)

I thought this will not break a lot, they just need to merge master, and github actions will triggered correctly. This project admin should remove the required checks for CircleCI.

**Comment by LittleFall** [comment]
This pr won't affect TiFlash, TiFlash only needs `.proto`s and ./generate_cpp.sh`.
And won't affect TiSpark.

**Comment by Xuanwo** [comment]
Let's make tipb great again! @breeswish @skyzh

**Comment by skyzh** [comment]
> Let's make tipb great again! @breeswish @skyzh

I don't have write permission to this repo. Maybe we should @breeswish

**Comment by breeswish** [comment]
Sadly I don't have permissions as well. Maybe @bb7133 can help.

**Pull request merged**
